### PR TITLE
Support Python 3.7

### PR DESCRIPTION
**Pull request by withsmilo**
* related issue : https://github.com/ucbrise/clipper/issues/580
* This PR will create new Docker images.
  * py37-rpc
  * mxnet37-container
  * pytorch37-container
  * tf37-container
  * pyspark37-container
  * python37-closure-container
  * keras37-container

### CODE REVIEW

File: clipper_admin/clipper_admin/deployers/keras.py
Comment (jacekwachowiak) [inline clipper_admin/clipper_admin/deployers/keras.py]: I think this should be `keras37`

Comment (withsmilo) [reply clipper_admin/clipper_admin/deployers/keras.py]: Oh! my mistake! I will fix it tonight.

**Comment by rkooo567** [comment]
CI jobs are pended a lot. The job for the last commit will be running from now on.

File: clipper_admin/clipper_admin/deployers/keras.py
Comment (withsmilo) [reply clipper_admin/clipper_admin/deployers/keras.py]: fixed!

**Comment by rkooo567** [comment]
jenkins test this please.

**Review by rkooo567 — DISMISSED** [review]
Looks good to me! I will merge if the last CI is passed.

**Comment by rkooo567** [comment]
@simon-mo I will wait for you to merge.

**Comment by simon-mo** [comment]
@withsmilo Thank you for the great work!

Can we:
- [ ] Build a python3.7 test container and test deploy_python_closure?
- [ ] In the doc, say that 3.7 support is still experimental (because we haven't tested the compatibility with other frameworks)

**Comment by withsmilo** [comment]
@simon-mo 
Current `py35tests` Docker image built by `ClipperPy35TestsDockerfile`, uses Python 3.6. We have two choices.

1. How about try to upgrade the Python version from 3.6 to 3.7 in `py35tests`? If we do this, we will use python 3.7 as the base version for unit testing. If we don't use a library of dependencies on Python 3.6, this approach will be no problem.

2. Let's conduct a temporary unit test with python 3.7.

\+ we need to rename `ClipperPy35TestsDockerfile` to `ClipperPy3TestsDockerfile`.

**Comment by withsmilo** [comment]
@simon-mo I will test to use python 3.7 as the base version for unit testing.

**Comment by withsmilo** [comment]
Jenkins test this please

**Comment by rkooo567** [comment]
@withsmilo Current CI is very unstable with this log. 

```
ReadTimeout: UnixHTTPConnectionPool(host='localhost', port=None): Read timed out. (read timeout=60) 
```

It can be one of our dependency issues like `urllib3` last time. I will run the `develop` on my vm and see if it is reproduced.

**Comment by rkooo567** [comment]
2 things.

1. Based on logs of job 2074, it seems like new python dev container is failed to be built. (So, I think it is independent to our main build failure)
2. I could success to pass test on my local environment for current `develop`, so the error might be Jenkins issue. We'll see if build keeps failing tomorrow.

**Comment by withsmilo** [comment]
@rkooo567 Wow, so unstable. :disappointed_relieved: Our Jenkins worker is working too hard now. (https://amplab.cs.berkeley.edu/jenkins/computer/amp-jenkins-staging-worker-02/builds)
I will run this PR this night again.

**Comment by withsmilo** [comment]
Jenkins test this please

**Comment by withsmilo** [comment]
Docker daemon might be killed on Jenkins worker. I’ll waiting for recovery. 

>`
failed to get digest sha256:fe3ef29c73f3ebaffa0dead8391b75be18894d771a841c28ca1140fec358c5e2: open /var/lib/docker/image/aufs/imagedb/content/sha256/fe3ef29c73f3ebaffa0dead8391b75be18894d771a841c28ca1140fec358c5e2: no such file or directory`

**Comment by rkooo567** [comment]
Jenkins test this please

**Comment by rkooo567** [comment]
@withsmilo @simon-mo The temporary test passes. I think we should decide what version of python we will test. What I am worried is that we will break python 3.6 or 3.5 because of features that are only supported by 3.7, but we don't know that happens because we only test with python 3.7. What do you guys think is the best idea for testing?

1. Use python 3.7 since most of new users will be use python 3.7.
2. Use python 3.6. Since we proved python 3.7 passes unit tests, it should be fine to keep developing without testing python 3.7 because it is backward compatible.
3. Test both. (I think it is not a good idea given our resources).

**Comment by withsmilo** [comment]
@rkooo567 @simon-mo I agree we can't keep all the Python versions for unit test. How about this?
* current unit test targets
  * python 2.7
  * python 3.6
* changed unit test targets on Jan. 1, 2020. (python 2.7 will be retired. ([link](https://pythonclock.org/)))
  * python 3.6
  * python 3.7

**Comment by withsmilo** [comment]
hmm.. I will add some retry routines to handle Docker's timeout errors.

**Comment by rkooo567** [comment]
@withsmilo I agree that we can stop supporting python 2.7 around the time python 2.7 retires. 👍 

Also, about CI, if retry doesn't work, we should probably investigate the problem like the last time (mxnet). If that happens, I will create a new issues!

**Comment by withsmilo** [comment]
Please do not merge yet. I will test to change unit test targets to python 3.6 and python 3.7.

**Comment by withsmilo** [comment]
@rkooo567 @simon-mo 
After Python 2.7 retired, we have to activate these lines to use Python 3.6 and 3.7 for unit test.
https://github.com/ucbrise/clipper/pull/723/files#diff-72ca7a5f8b86c887fd5695574f26fce0R65

**Comment by simon-mo** [comment]
@withsmilo After 2020, I'm fine with even just testing with 3.6. and only run a small subset of the tests in other python version if only necessary. bascially:

- python3.6: full set
- python3.5, python 3.7: only deploy python closure.

**Review by simon-mo — APPROVED** [review]

**Comment by withsmilo** [comment]
@simon-mo Good idea!

**Pull request merged**
